CRM/Civioffice/DocumentRendererType.php

```diff
@@ -190,64 +184,60 @@ public function replaceAllTokens($string, $token_contexts = []): string
-            switch ($entity_type) {
-                case 'civioffice':
-                    $token_row->context('civioffice.live_snippets', $context['live_snippets']);
-                    break;
-                case 'contact':
-                    $token_row->context('contactId', $context['entity_id']);
-
-                    /*
-                     * FIXME: Unfortunately we get &lt; and &gt; from civi backend so we need to decode them back to < and > with htmlspecialchars_decode()
-                     * This is postponed as it might be risky as it either breaks xml or leads to further problems
-                     * https://github.com/systopia/de.systopia.civioffice/issues/3
-                     */
-
-                    break;
-                case 'contribution':
-                    $token_row->context('contributionId', $context['entity_id']);
-                    $token_row->context('contribution', $context['entity']);
-                    break;
-                case 'participant':
-                    $token_row->context('participantId', $context['entity_id']);
-                    $token_row->context('participant', $context['entity']);
-                    break;
-                case 'event':
-                    $token_row->context('eventId', $context['entity_id']);
-                    $token_row->context('event', $context['entity']);
-                    break;
-                case 'membership':
-                    $token_row->context('membershipId', $context['entity_id']);
-                    $token_row->context('membership', $context['entity']);
-                    break;
-                default:
-                    // todo: implement?
-                    throw new Exception('replaceAllTokens not implemented for entity ' . $entity_type);
-                    break;
+            if (!isset($entity_token_context[$entity_type])) {
+                throw new Exception(
+                    E::ts('Could not determine token context for entity type %1.', [1 => $entity_type])
+                );
             }
+            $token_contexts_schema[$entity_token_context[$entity_type]] = $context['entity_id'];
         }
 
-        $processor->evaluate();
-        return $token_row->render($identifier);
+        $this->tokenProcessor->addSchema(array_keys($token_contexts_schema));
+        $token_row = $this->tokenProcessor->addRow($token_contexts_schema)
+            ->format('text/html');
+
+        // Replace tokens in Live Snippets and update token contexts.
+        $this->replaceLiveSnippetTokens($token_row);
+        $token_contexts_schema['civioffice.live_snippets'] = $this->liveSnippets;
+        $token_row->context($token_contexts_schema);
+
+        return $token_row;
     }
 
-    /*
-     * Could be used to convert larger batches of strings and/or contact ids
+    /**
+     * Builds a mapping of entity type names and their corresponding token context schema
```

Comment:
I think it's actually not a schema, but the context key for the entity ID (at least as of now).

Reply:
The method name for retrieving the token context "key" is called `getEntityContextSchema()` in [CRM_Core_TokenTrait](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/TokenTrait.php). I'll just call them *token context schema identifiers*.